src/test/java/gov/healthit/chpl/aqa/stepDefinitions/Base.java

```diff
@@ -95,6 +95,38 @@ public void checkIfFileIsDownloaded(final String fileName)throws FileNotFoundExc
         }
     }
 
+    /**
+     * Checks if the file contains all the data and is downloaded completely or not.
+     * @param fileName expected downloaded file name
+     * @throws FileNotFoundException if the expected file not found
+     */
+    public void checkCompleteFileDownload(final String fileName) throws FileNotFoundException {
+        boolean completeFileDwnld = true;
+        int retryCount = 0;
+        Long size = new Long(-1);
```

Comment:
What is this used for? It gets set, and checked against, but is never used

Reply:
It's been used to make sure whether the complete file is downloaded

Reply:
Where? What line?

Reply:
Refer line number : 111
Here is the code snippet:
 Long downloadFileSize = file.length();
                    if (size < downloadFileSize) {
                        size = downloadFileSize;
                    } else {
                        completeFileDwnld = false;
                        break;
                    }

Reply:
I'm trying to follow this logic. In pseudocode, what you're doing is:

```
while I haven't completed the download, and haven't reached the max retry value:
  for each file in the download directory:
    get the size of the file
    if the size of the last file is >= the new file's size
      stop looking at the files, and say we've found our file
    otherwise
      wait for 5 seconds and check the next file
If I havnen't said we found the file
  report that it wasn't found
```

* Why pass in the name of the file if you're not checking it against the file that was found?
* What would happen if there was more than one file in this directory?
* Why is `completeFileDwnld` true at the start of the method, and false at the end, assuming everything worked correctly? It seems to me that "Complete File Downloaded" should be true at the end, not the beginning

Reply:
Trying to follow this as well.. maybe if you changed the while condition to "!completeFileDownloaded and retryCount <= MAX_RETRY_COUNT" then the booleans could be switched to make more sense as Andrew described.

Reply:
In the download file folder, what would happen if there were two files there? Let's say there's already a file called "chpl-2015.xml" with a size of 100Mb, and we start downloading the sed-details file. Assuming it doesn't download immediately, it's name would be something like "sed-details.crdownload", and let's assume it's going to be 30Kb, but gets 5Kb every 5 seconds. Here's what it looks like on each run through the for loop

Assuming chpl-2015 is first
```
+-----+----------------+--------+---------------------------------+-------------------------------------------+
|size |downloadFileSize|download|file name                        |State                                      |
+-----+----------------+--------+---------------------------------+-------------------------------------------+
|-1   |undefined       |false   |undefined                        |initial state                              |
+-----+----------------+--------+---------------------------------+-------------------------------------------+
|100Mb|100Mb           |false   |chpl-2015.xml                    |first file check, now sleep for 5 sec      |
+-----+----------------+--------+---------------------------------+-------------------------------------------+
|100Mb|10Kb            |true    |sed-details.crdownload           |"Found" file, but it's the wrong one       |
+-----+----------------+--------+---------------------------------+-------------------------------------------+
```

Assuming sed-details is first

```
+-----+----------------+--------+---------------------------------+--------------------------------------------+
|size |downloadFileSize|download|file name                        |State                                       |
+-----+----------------+--------+---------------------------------+--------------------------------------------+
|-1   |undefined       |false   |undefined                        |initial state                               |
+-----+----------------+--------+---------------------------------+--------------------------------------------+
|5Kb  |5Kb             |false   |sed-details.crdownload           |first file check, now sleep for 5 sec       |
+-----+----------------+--------+---------------------------------+--------------------------------------------+
|100Mb|100Mb           |false   |chpl-2015.xml                    |Second file check, now sleep for 5 more sec |
+-----+----------------+--------+---------------------------------+--------------------------------------------+
|100Mb|15Kb            |true    |sed-details.crdownload           |Found it again, but also the wrong one      |
+-----+----------------+--------+---------------------------------+--------------------------------------------+
```

Reply:
@andlar There won't be more than one file in the directory because the AND condition (clears the directory) is executed before this step so there is just one file in the download directory

Reply:
The actual logic that I have used is like:- 
while I haven't completed the download, and haven't reached the max retry value:
  for each file in the download directory: 
    get the size of the file
  if the size of the current file is > size variable
	Sleep to allow the file to be downloaded completely that doesn't allow the browser to close
    Otherwise
          Size variable = current file size
    Then the file is completely downloaded
     Else If
	After going through all the retry counts are and still, the file is not downloaded 
     Throw exception
	File not found

Reply:
AND code snippet that I'm reusing:-
    /**
     * Print filenames if directory is not empty.
     * Assert that download directory is empty
     * @throws IOException if unable to clean the directory
     */
    @And("^the download directory is empty$")
    public void checkifDownloadDirectoryisEmpty() throws IOException {
        System.out.println("Clearing directory: " + Hooks.getDownloadDirectory().getAbsolutePath());
        for (String fileName : Hooks.getDownloadDirectory().list()) {
            System.out.println("Removing: " + fileName);
        }
        FileUtils.cleanDirectory(Hooks.getDownloadDirectory());
        assertFalse(Hooks.getDownloadDirectory().list().length > 0, "directory is not empty");
    }